### PR TITLE
Add references to Nbshell in export templates, reports and scripts

### LINKED ISSUE #15014: Documentation is unclear on how to access data structures in Scripts/Reports/Config Templates/Export Templates
### Change Type

Addition

### Area

Customization

### Proposed Changes

Hi!

First of all, sorry is this is my falt for not seeing it correctly.

The issue is, I'm starting to dive into the customization/scripting capabilities of Netbox, like Reports and Config Templates, and I'm having a very hard time knowing how to access the data from Netbox itself, since the documentation only provides a very simple example. I don't know which "objects" are accesible from a config template, nor I know what fields those objects have or the information they contain, or the methods they have.

For example: In a config template for a router that has several circuits connected, I want to know the bandwidth of each circuit in order to configure QoS in the router. How do I do that? I know (and I only know it from the example) that I can access "device.interfaces.all()". And from there? What does the data structure for "interface" look like? What other fields are there in the "device" object besides the obvious name, type, role... etc? What other methods, besides "all()" are there. Is there a filter? What's the syntax for it?

There are a lot of unanswered questions. If this information can be found somewhere, even if it's buried in the source code, I would be happy to contribute to the documentation page by summarizing it, so that newcomers don't get frustrated trying to find it. I just need somebody to tell me where to look.

Thanks!!

### PR DESCRIPTION
**Pull request by 8ctorres**
### Fixes: #15014

The Netbox shell is a very powerful resource for people trying to develop their own scripts, reports or configuration templates, but I feel it is difficult to find in the documentation. For this reason, I have added a few references to it in the pages, so that a new user can find it much sooner than it took me to find.

### CODE REVIEW

File: docs/administration/netbox-shell.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # The NetBox Python Shell
 
-NetBox includes a Python management shell within which objects can be directly queried, created, modified, and deleted. To enter the shell, run the following command:
+NetBox includes a Python management shell within which objects can be directly queried, created, modified, and deleted. To enter the shell, run the following command from a shell that has the netbox virtualenv activated:
```
Comment (jeremystretch) [inline docs/administration/netbox-shell.md:3]:
```suggestion
NetBox includes a Python management shell within which objects can be directly queried, created, modified, and deleted. To enter the shell, run the following command with the appropriate Python virtual environment activate:
```

File: docs/administration/netbox-shell.md [hunk 2]
```diff
@@ -143,6 +143,40 @@ To return the inverse of a filtered queryset, use `exclude()` instead of `filter
...
+
+```
+>>> site.circuit_terminations.all()
+<RestrictedQuerySet [<CircuitTermination: 20899518: Termination A>, <CircuitTermination: DT00018356: Termination A>]>
+>>>
+```
+
+The same methods (all, filter, exclude, get...) can be used in this kind of queries.
+
+
```
Comment (jeremystretch) [inline docs/administration/netbox-shell.md:179]: This is out of scope for the NetBox documentation as it describes generic behavior of the [Django ORM](https://docs.djangoproject.com/en/stable/topics/db/queries/). Rather than duplicate Django documentation, we should refer users there to read further.

File: docs/customization/custom-scripts.md [hunk 3]
```diff
@@ -56,6 +56,15 @@ class AnotherCustomScript(Script):
 script_order = (MyCustomScript, AnotherCustomScript)
 ```
 
+### The run() method
```
Comment (jeremystretch) [inline docs/customization/custom-scripts.md:59]: Always use markup to highlight code excerpts, and capitalize heading titles.
```suggestion
### The `run()` Method
```

File: docs/customization/custom-scripts.md [hunk 4]
```diff
@@ -56,6 +56,15 @@ class AnotherCustomScript(Script):
...
+
+The run() method is the entrypoint for the script, and it runs in the context of Netbox's own execution environment. This means from here, everything inside Netbox itself is accesible. The [Netbox Shell](../administration/netbox-shell.md) is a good resource to keep in hand, since it allows to see the objects in Netbox in the same way the run() method of the script does.
```
Comment (jeremystretch) [inline docs/customization/custom-scripts.md:61]: NetBox is always written as either "NetBox" (proper usage) or "netbox" (referencing code/resources); never "Netbox."

File: docs/customization/export-templates.md [hunk 5]
```diff
@@ -32,6 +32,8 @@ If you need to use the config context data in an export template, you'll should
 {% endfor %}
 ```
 
+To see all the attributes of a given object, you can use the [Netbox Shell](../administration/netbox-shell.md). It supports autocompletion and allows one to see all of the methods and attributes a given object type has. All of them can be called from within a Jinja template. Using queries from one object to another, one can navigate pretty much the entire Netbox object model. For instance, from an export template for sites, one can get the devices that are in that site, the circuits that are connected to those devices, the providers that serve those circuits... etc, so an export template is not limited to just the model that it's being called from. In fact, the same result can be achieved in different ways, depending on which model you start from.
```
Comment (jeremystretch) [inline docs/customization/export-templates.md:35]: This should be reduced to a "tip" admonition as the topic strays from export templates. Just point out that object attributes can be explored using `nbshell`.

File: docs/customization/reports.md [hunk 6]
```diff
@@ -29,7 +29,9 @@ class DeviceIPsReport(Report):
     description = "Check that every device has a primary IP address assigned"
 ```
 
-Within each report class, we'll create a number of test methods to execute our report's logic. In DeviceConnectionsReport, for instance, we want to ensure that every live device has a console connection, an out-of-band management connection, and two power connections.
+Within each report class, we'll create a number of test methods to execute our report's logic. The method's name must start with "test_" and it takes no arguments.
```
Comment (jeremystretch) [inline docs/customization/reports.md:32]:
```suggestion
Within each report class, we'll create a number of test methods to execute our report's logic. The name of each method must begin with `test_` and the method must accept no arguments.
```

File: docs/customization/reports.md [hunk 7]
```diff
@@ -82,6 +84,8 @@ class DeviceConnectionsReport(Report):
 
 As you can see, reports are completely customizable. Validation logic can be as simple or as complex as needed. Also note that the `description` attribute support markdown syntax. It will be rendered in the report list page.
 
+In the same way scripts do, reports run from within Netbox's own environment and can access the objects inside Netbox directly. The [Netbox Shell](../administration/netbox-shell.md) is a good resource to keep in hand, since it allows to see the objects in Netbox in the same way the test methods of a report do.
```
Comment (jeremystretch) [inline docs/customization/reports.md:87]: Please correct "Netbox" to NetBox. This should also probably render as a tip admonition.

File: docs/features/configuration-rendering.md [hunk 8]
```diff
@@ -35,6 +35,8 @@ Configuration templates are written in the [Jinja2 templating language](https://
 {% endblock %}
 ```
 
+To see all the attributes of a given object, you can use the [Netbox Shell](../administration/netbox-shell.md). It supports autocompletion and allows one to see all of the methods and attributes a given object type has. All of them can be called from within the configuration template. Also, other objects, for instance, a device's interfaces or connected circuits can be accessed from the template itself, so the rendered configuration may include information not only about the device itself, but also related objects like IP addresses or circuits.
```
Comment (jeremystretch) [inline docs/features/configuration-rendering.md:38]: As with export templates, lets compress this into a brief tip admonition pointing the reader to the `nbshell` page.

**Review by jeremystretch — CHANGES_REQUESTED** [review]
Thanks @8ctorres. Let's compress some of the commentary on using `nbshell` to ensure we don't disrupt from the primary flow of each page.

**Comment by 8ctorres** [comment]
Hi @jeremystretch thanks for your feedback. I will try to make some changes this weekend and update the PR.

**Comment by jeremystretch** [comment]
@8ctorres checking in. Are you still working on this PR?

**Comment by 8ctorres** [comment]
Hi @jeremystretch thanks for checking. I haven't really had the time these past few weeks, I apologize for the delay. I will try to get this done before the end of the month. Sorry again

File: docs/administration/netbox-shell.md (hunk 2 above)
Comment (8ctorres) [reply docs/administration/netbox-shell.md]: Done. I reduced it to just one line about the get() method, and left the tip admonition pointing users to the Django documentation page

File: docs/customization/custom-scripts.md (hunk 4 above)
Comment (8ctorres) [reply docs/customization/custom-scripts.md]: Done. Sorry for the mistake

File: docs/customization/reports.md (hunk 7 above)
Comment (8ctorres) [reply docs/customization/reports.md]: Done

File: docs/features/configuration-rendering.md (hunk 8 above)
Comment (8ctorres) [reply docs/features/configuration-rendering.md]: Done

**Comment by 8ctorres** [comment]
Hi @jeremystretch , sorry again for the delay with this. I have reviewed your suggestions and added another commit with a few changes. About reports, I see that they're being deprecated in v4, so maybe the documentation on that section isn't going to be so relevant anymore.

Thanks!

**Comment by arthanson** [comment]
@8ctorres can you please fix the merge conflicts

**Comment by 8ctorres** [comment]
Hi @arthanson sorry for the delay. I fixed the merge conflicts in the custom scripts page. For the custom reports page, since they're being deprecated anyways, I just rescinded all my changes and kept the incoming ones. I hope that's fine

**Comment by arthanson** [comment]
@8ctorres Looks good, can you remove the changes in netbox/translations/en/LC_MESSAGES/django.po not sure why or how they got there

**Comment by 8ctorres** [comment]
Hi @arthanson, I think it must be some automatic github action that updates translation/locale metadata. In any case, I just reverted that commit so it should be good now.

File: docs/administration/netbox-shell.md [hunk 9]
```diff
@@ -105,6 +105,8 @@ This approach can span multiple levels of relations. For example, the following
 >>> IPAddress.objects.filter(interface__device__site__region__slug="north-america")
 ```
 
+The `get()` method can be used in place of the `filter()` method if we expect this query to return only one object. This method will yield the actual object resulting from the query, instead of a QuerySet. If the query returns more than one result, it will raise an exception.
```
Comment (jeremystretch) [inline docs/administration/netbox-shell.md:108]: This is redundant to information provided in the [Django documentation](https://docs.djangoproject.com/en/5.0/ref/models/querysets/). Please reference it here instead of duplicating text that we'll need to then maintain.

File: docs/administration/netbox-shell.md [hunk 10]
```diff
@@ -143,6 +145,8 @@ To return the inverse of a filtered queryset, use `exclude()` instead of `filter
 346
 ```
 
+Queries can all also be executed from a particular object instead of from the model itself. For instance, to get all circuits that are assigned to one site, it is easier to filter from the site itself, instead of using the "Circuit" model and building the query from there. This is particularly useful for configuration templates and export templates, since it allows to query other database objects that are related to the object that we're rendering the template for. The same methods (all, filter, exclude, get...) can be used in this kind of queries.
```
Comment (jeremystretch) [inline docs/administration/netbox-shell.md:148]: Again, this is redundant to the Django documentation, which provides plenty of context and examples.

File: docs/customization/custom-scripts.md [hunk 11]
```diff
@@ -65,6 +65,18 @@ class AnotherCustomScript(Script):
 script_order = (MyCustomScript, AnotherCustomScript)
 ```
 
+The `run()` method is the entrypoint for the script, and it runs in the context of NetBox's own execution environment. This means from here, everything inside NetBox itself is accesible. The [NetBox Shell](../administration/netbox-shell.md) is a good resource to keep in hand, since it allows to see the objects in NetBox in the same way the script does.
```
Comment (jeremystretch) [inline docs/customization/custom-scripts.md:68]: We've already introduced the `run()` method above; this makes no sense to put here. The link to the `nbshell` documentation should be a tip admonition as it doesn't directly relate to authoring a script.

File: docs/customization/custom-scripts.md [hunk 12]
```diff
@@ -65,6 +65,18 @@ class AnotherCustomScript(Script):
...
+The `run()` method can itself call other methods that are in the same module but outside the "MyCustomScript" class, and if there are several scripts in the same module (this is, in the same Python file), both scripts can reuse the same auxiliary methods, keeping the code cleaner. For this reason, it is encouraged to keep similar scripts in the same module. Script can return a string, which will be displayed in a text box in the web interface after the script finishes. This is useful, for instance, for returning a piece of configuration or information that you want the user to be able to easily copy and paste somewhere else.
+
+
+## Module Attributes
+
+### `name`
+
+You can define `name` within a script module (the Python file which contains one or more scripts) to set the module name. If `name` is not defined, the module's file name will be used.
+
+
```
Comment (jeremystretch) [inline docs/customization/custom-scripts.md:79]: This is no longer supported.
```suggestion
```

**Review by jeremystretch — CHANGES_REQUESTED** [review]
@8ctorres could you please rebase your changes against the current `develop` branch and remove the unrelated commits? (I tired to myself but you have not given access for PR reviewers to push to your branch.)

**Comment by 8ctorres** [comment]
Hi @jeremystretch I rebased my branch onto the current develop branch, and reverted the changes to the "netbox/translations/en/LC_MESSAGES/django.po" (it's an automatic action, but still). Also implemented the changes you suggested.

I think it should be good now. Thanks for your patience.

**Comment by jeremystretch** [comment]
@8ctorres it looks like something has gone wrong with this PR, as it includes a lot of changes that aren't in-scope for the issue. Please see if you can rebase your branch, or start a new one, and resubmit this PR.
